### PR TITLE
Fix converge issue in Orca TF Estimator

### PR DESCRIPTION
**Pull request by jenniew**
Fix converge issue in Orca TF Estimator.
Related issue: https://github.com/intel-analytics/analytics-zoo/issues/3081

### CODE REVIEW

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 1]
```diff
@@ -628,6 +631,8 @@ def fit(self, data,
                              hard_code_batch_size=hard_code_batch_size,
                              sequential_order=False, shuffle=True,
                              auto_shard_files=auto_shard_files)
+        if isinstance(dataset, TFNdarrayDataset):
+            dataset = _standarize_feature_label_dataset(dataset, self.model.model)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:635]: It seems that we are doing this for `KerasModel`, but not `TFOptimizer`? Why don't we handle it in `TFNdarrayDataset`? @yangw1234

Comment (yangw1234) [reply pyzoo/zoo/orca/learn/tf/estimator.py:635]: This is a special treatment for Keras models because Keras requires a target to have a shape of [batch, 1] not [batch,]. This is not generally the case for other models.

Comment (jason-dai) [reply pyzoo/zoo/orca/learn/tf/estimator.py:635]: > This is a special treatment for Keras models because Keras requires a target to have a shape of [batch, 1] not [batch,]. This is not generally the case for other models.

It seems that the processing is not very consistent.
1) It's not handled properly in `KerasModel.evaluate` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L209) or `KerasModel.predict` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L275)
2) It's not handled in `TFOptimizer.from_keras` at all

I think we need to handle it properly in TFPark (both `KerasModel` and `TFOptimizer`), if not in `TFNdarrayDataset`

Comment (yangw1234) [reply pyzoo/zoo/orca/learn/tf/estimator.py:635]: Yeah, we should handle it properly in both `KerasModel` and `TFOptimizer`!

Comment (yangw1234) [reply pyzoo/zoo/orca/learn/tf/estimator.py:635]: @jason-dai While we are on this subject, `TFOptimizer.from_keras` is essentially trying to reimplement tf.keras.fit in analytics zoo. 

I think it is very hard, if not impossible, to match the exact behaviors of keras in analytics zoo. There are more than 8k lines of code in tf.keras just dealing with the training logic and more than 5k lines of code of tests covering it. In contrast, tf_optimizer.py only has 761 lines of code. I think there must be some logic we are missing.

So I am thinking, should we support all keras models (both tf 1.15 and 2.x) in ray-based estimators? And use bigdl based estimator for graph API only. It is much easier to just call keras instead of reimplementing it.  And other features like callbacks are supported out of the box.

Comment (jason-dai) [reply pyzoo/zoo/orca/learn/tf/estimator.py:635]: > @jason-dai While we are on this subject, `TFOptimizer.from_keras` is essentially trying to reimplement tf.keras.fit in analytics zoo.
> 
> I think it is very hard, if not impossible, to match the exact behaviors of keras in analytics zoo. There are more than 8k lines of code in tf.keras just dealing with the training logic and more than 5k lines of code of tests covering it. In contrast, tf_optimizer.py only has 761 lines of code. I think there must be some logic we are missing.
> 
> So I am thinking, should we support all keras models (both tf 1.15 and 2.x) in ray-based estimators? And use bigdl based estimator for graph API only. It is much easier to just call keras instead of reimplementing it. And other features like callbacks are supported out of the box.

That is orthogonal to this PR and we can handle it in a separate issue.

Looks like we are using `KerasModel` in both `TFKerasWrapper.predict` and `TFKerasWrapper.evaluate`, I wonder why we are not using `KerasModel` in `TFKerasWrapper.fit`? if so, we can just make sure `KerasModel` has the correct behavior.

**Comment by yangw1234** [comment]
I still cannot figure out why this fix works. 
Is it possible to write a unit test for it?

**Comment by jason-dai** [comment]
> I still cannot figure out why this fix works.
> Is it possible to write a unit test for it?

That's exactly what you did in `KerasModel.fit` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L142), yes? BTW, why didn't we use `KerasModel.fit` in `TFKerasWrapper.fit`?

**Comment by yangw1234** [comment]
> > I still cannot figure out why this fix works.
> > Is it possible to write a unit test for it?
> 
> That's exactly what you did in `KerasModel.fit` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L142), yes? BTW, why didn't we use `KerasModel.fit` in `TFKerasWrapper.fit`?

I mean the standardize_... func just add one dimension to label‘s shape, it should not affect convergence. 

I am not sure why Model.fit is not used here. Maybe to support checkpoint I guess?

**Comment by jenniew** [comment]
> > > I still cannot figure out why this fix works.
> > > Is it possible to write a unit test for it?
> > 
> > 
> > That's exactly what you did in `KerasModel.fit` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L142), yes? BTW, why didn't we use `KerasModel.fit` in `TFKerasWrapper.fit`?
> 
> I mean the standardize_... func just add one dimension to label‘s shape, it should not affect convergence.
> 
> I am not sure why Model.fit is not used here. Maybe to support checkpoint I guess?

Yes, KerasModel does not support load bigdl checkpoint up to now. So I use low-level TFOptimizer to support that. If we add support to load bigdl checkpoint, then we can use KerasModel.fit directly.

**Comment by jason-dai** [comment]
> > > > I still cannot figure out why this fix works.
> > > > Is it possible to write a unit test for it?
> > > 
> > > 
> > > That's exactly what you did in `KerasModel.fit` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L142), yes? BTW, why didn't we use `KerasModel.fit` in `TFKerasWrapper.fit`?
> > 
> > 
> > I mean the standardize_... func just add one dimension to label‘s shape, it should not affect convergence.
> > I am not sure why Model.fit is not used here. Maybe to support checkpoint I guess?
> 
> Yes, KerasModel does not support load bigdl checkpoint up to now. So I use low-level TFOptimizer to support that. If we add support to load bigdl checkpoint in KerasModel, then we can use KerasModel.fit directly.

OK - I think we can merge this PR for now (maybe also fixing [`KerasModel.evaluate`](https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L209) an [`KerasModel.predict`](https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L275)); we can add checkpoint support in `KerasModel` later.

**Comment by jenniew** [comment]
jenkins pass: http://10.239.47.210:18888/job/ZOO-PR-Validation/4573/

**Comment by jason-dai** [comment]
@jenniew you may also want to fix `KerasModel.evaluate` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L209) and `KerasModel.predict` (https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/tfpark/model.py#L275)
